OpticalFlow : Add Onboard Optical Flow for Bebop

jberaud opened the pull request:
Here is a pull request to support the integrated camera sensor of the bebop and use it as an optical flow sensor using the algorithm found in the PX4 optical flow code https://github.com/PX4/Flow/blob/master/src/modules/flow/flow.c

I had to remove ARM assembly code which I will re-introduce later under another form, i.e probably a separate assembly file implementing functions for a fixed size.

There is a possibility to record the raw video stream in a file, along with the gyro datas, which can be very useful to process off-board and compare. I have written a small linux program that implements an optical flow on a video file.

This PR adds support for:
-> A v4l2 capture driver that is generic and makes it possible to capture frames, set format and crop.
-> A generic optical flow driver that uses this driver
-> A driver for aptina mt9v117 camera sensor over i2c (for config) that only supports qvga and max framerate for now but can be modified
-> A new HAL interface : OpticalFlow - I haven't found a more elegant way to implement this function but I am totally ok to implement one if someone has a more elegant idea
-> A new AP_OpticalFlow driver that uses this HAL interface to get datas and gives a functor cb to get gyro angle datas from the AHRS.

It has been tested by looking at the dataflash logs and comparing between gyro datas and calculated flow angles and it looks ok. There may be some upcoming fixes when the flight tests are done but this will require the integration of the sonar first.

This Pull Request is a bit big but there were some missing functionalities, code integration, interface changes and some fixes to make it work.

Review is of course highly appreciated along with comments and tests.

lucasdemarchi commented:
Hi @jberaud nice to see this being implemented. It will be very nice to have it in the same board as the flight stack.

I guess it should be noted that the camera should be placed in the same position (relative to the center of gravity) as the camera. Or is it not necessary?

I'm going to add some comments to the fixes and refactors you did in the current code. I think the PR is big enough and currently I can't review it entirely. It would be nice if you could send the fixes/refactors separately (even if they still are in this PR). This way we can apply them while we discuss the feature you are adding.

lucasdemarchi commented:
I think I reviewed everything touching the current classes. I need to stop here and come back to the actual flow implementation later. As I said, opening another PR for these commits would be nice.

jberaud commented:
>I guess it should be noted that the camera should be placed in the same position (relative to the center of gravity) as the camera. Or is it not necessary?

The camera needs to be placed in the same orientation. Regarding the position, @priseborough will have a better understanding than I. I intuitively think that it is better for it to be placed at the roll/pitch center of the drone but there is nothing specific to the bebop here. It also needs to be placed as close as possible to the RangeFinder.

jberaud commented:
>I think I reviewed everything touching the current classes. I need to stop here and come back to the actual flow implementation later. As I said, opening another PR for these commits would be nice.

I will open another PR for the GPIO, I2c, and PWM changes with fixes for the issues you pointed out. I will wait for the rest of this PR to be reviewed to post another one for the core optical flow.

jberaud commented:
Rebased on master. PWM, GPIO, I2CDriver have been pushed the mentioned PR

priseborough commented:
Regarding optical flow camera pose and position:

1) The camera Z axis should be aligned with the drone Z axis
2) The camera yaw rotation relative to the drone can be compensated for using the FLOW_ORIENT parameter
2) The camera (and range finder) should be located within 20cm of the centre of the drone

It would be nice to be able to define the position of every sensor relative to the drone centre using parameters

jberaud commented:
Thanks ! I had only tested with waf since it is what is used on travis. I have updated the code.

lucasdemarchi commented:
@jberaud sorry for the delay... lots of things to do before going on vacations.

Instead of doing a review here I decided to push a reviewed branch since I had a flaky internet connection and I think this would speed up the process to get this applied: https://github.com/lucasdemarchi/ardupilot/tree/review/bebop-optflow

Please take a look on the commits I added. This is not tested, but most of the changes are to coding style and to ease maintenance of this code. Please make note of these changes for next PRs.

I don't know if it's my setup here (I recently changed my computer), but this PR was not even compiling for bebop. There were some missing casts. See last commit. I fixed it by adding the casts. Please also add bebop to travis, so we catch these problems early on.

As for the implementation. I see you are using a separate thread to read the frames. I don't think 13 is a good priority value. This is above the main thread. I'm also of the opinion the main thread priority is already wrong, but this is another topic. Have you measured any increase in scheduler misses when running with onboard flow?

jberaud commented:
>Please take a look on the commits I added. This is not tested, but most of the changes are to coding style and to ease maintenance of this code. Please make note of these changes for next PRs.

Thanks for that. I will apply these rules in the next PRs I will make. Some of them are just things I have let pass, others are real rules I haven't applied due to my not knowing them.
Anyway I will follow these rules. Could we have a code checker that automatizes this process? Just to be easier and not bother you with just failure to apply these rules?

>I don't know if it's my setup here (I recently changed my computer), but this PR was not even compiling for bebop. There were some missing casts. See last commit. I fixed it by adding the casts. Please also add bebop to travis, so we catch these problems early on.

It was obviously compiling on my setup, which is using the regular Makefile. Maybe we should just fix the makefile to add warning checks. I will add that to travis, maybe using waf, what do you think?

>As for the implementation. I see you are using a separate thread to read the frames. I don't think 13 is a good priority value. This is above the main thread. I'm also of the opinion the main thread priority is already wrong, but this is another topic. Have you measured any increase in scheduler misses when running with onboard flow?

There is absolutely no noticeable increase in the misses with these flow dimensions. It is not the same with bigger frames. I have just tested your patches, along with a decrease to 11 for this priority and it works fine. I will modify my PR accordingly, also adding your corrections.

lucasdemarchi commented:
> Could we have a code checker that automatizes this process? Just to be easier and not bother you with just failure to apply these rules?

I'd be happy to have such a tool. PX4 has something for it, but at least their implementation needs all the codebase to be uniform. Ours is not.  If you have any idea of tool to do the check, let me know.

> It was obviously compiling on my setup, which is using the regular Makefile. Maybe we should just fix the makefile to add warning checks.

Here I was using regular Makefile as well since the waf build doesn't compile bebop yet.

> I will add the bebop build to travis, maybe using waf, what do you think?

Adding to travis should be a oneliner to Tools/scritps/build_all_travis.sh. For waf @guludo was working on it, but I think that since the board-inheritance patches to waf is already applied, supporting other linux boards with it should be very simple.

> There is absolutely no noticeable increase in the misses with these flow dimensions. It is not the same with bigger frames. I have just tested your patches, along with a decrease to 11 for this priority and it works fine. I will modify my PR accordingly, also adding your corrections.

Great. Thanks.

jberaud commented:
>I'd be happy to have such a tool. PX4 has something for it, but at least their implementation needs all the codebase to be uniform. Ours is not. If you have any idea of tool to do the check, let me know.

I'll ask and look around.

>Here I was using regular Makefile as well since the waf build doesn't compile bebop yet.

Then that's strange. Were you using the linaro toolchain that is indicated in the wiki? I didn't even have a warning about that because obviously I don't submit PRs with warnings.

ricardotk commented:
Hello @jberaud,

It's good to see this kind of things being implemented here, good job :+1: 

I've got this PR to run in a MinnowBoard and a simple web camera, since I don't have a bebop here (:

It's running well, I just had to do some minor fixes and change the V4L2 format from YUYV to NV12 (the web camera I'm using only supports YUYV and MJPG) since `compute_subpixel()` was crashing with a different format.

I'm not sure if the results I'm receiving from `compute_flow()` are correct, do you have an idea on how can I check this? If possible, send me the expected values or the values you've got during your tests.

Thanks!

jberaud commented:
> I've got this PR to run in a MinnowBoard and a simple web camera, since I don't have a bebop here (:

Good to see that running on another board, I have written this code to be as generic as possible.
>It's running well, I just had to do some minor fixes and change the V4L2 format from YUYV to NV12 (the web camera I'm using only supports YUYV and MJPG) since compute_subpixel() was crashing with a different format.

Basically it has been written to work on a planar format, which means it won't work with YUYV. In case you want to do that, you would have to rewrite all the pixel processing functions to increment by 2 pixels instead of one (only the Y terms are relevant to be used by this algorithm). What exactly did you do to change the format from YUYV to NV12?

>I'm not sure if the results I'm receiving from compute_flow() are correct, do you have an idea on how can I check this? If possible, send me the expected values or the values you've got during your tests.

In order to check that everything is correct, you can take your board and roll it back and forth for a few seconds. Then take the dataflash log and analyze them, for instance using MAVExplorer. graph OF.flowX and OF.bodyX and they should approximately be the same.

ricardotk commented:
>Basically it has been written to work on a planar format, which means it won't work with YUYV. In case you want to do that, you would have to rewrite all the pixel processing functions to increment by 2 pixels instead of one (only the Y terms are relevant to be used by this algorithm). What exactly did you do to change the format from YUYV to NV12?

I'm simply using [libv4l-convert] (http://linuxtv.org/downloads/v4l-dvb-apis/libv4l.html#id-1.4.9.2.5) to convert from one format to another,  it was pretty straightforward. I don't know how much overhead it's going to add add but I think rewriting all the pixel processing functions would be a better way to go.

>In order to check that everything is correct, you can take your board and roll it back and forth for a few seconds. Then take the dataflash log and analyze them, for instance using MAVExplorer. graph OF.flowX and OF.bodyX and they should approximately be the same.

Cool, I'll try it, thanks! (:

ricardotk commented on libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp:
```diff
@@ -0,0 +1,251 @@
+        AP_HAL::panic("OpticalFlow_Onboard: planar or monochrome format needed");
+    }
+
+    if (!_videoin->set_crop(left, top, crop_width, crop_height)) {
```
I tried to run without this `set_crop()`, even changing `HAL_OPTFLOW_ONBOARD_SENSOR_WIDTH` and `HAL_OPTFLOW_ONBOARD_CROP_WIDTH` it keeps crashing on `Flow_PX4::compute_subpixel()`, seems to be an "array index out of bounds" error. It solves when we hardcode change `_bytesperline` value. Why is this crop necessary? Shouldn't the code work even without this crop?

ricardotk replied:
In other words, is this supposed to work only in "squared" images? e.g. 240x240

jberaud replied:
Ricardo,
I suggest we continue this discussion on google groups if you don't mind. I don't want to give the impression that this code is under development because it is functional on the Bebop.
We'll then sort out how to make it work on your board.

ricardotk replied:
@jberaud makes sense, lets talk in ardupilot's gitter group then.

rmackay9 commented:
from my point of view, I'm ok for this to go in.  I flew it on my Bebop and it was fine.
I asked Tridge if he was ok with the hal layer addition of video and camera and he seems happy with the concept.
If there are no further questions/objections maybe i'll push to master soon...

jberaud commented:
That would be great. Thanks Randy.

tridge commented:
I've reviewed this and it is a really great piece of work. Congratulations Julien, and many thanks to Lucas for the detailed review and updated.
This is a really significant milestone, thanks very much!
applied
